apollo-federation/src/sources/connect/json_selection/selection_trie.rs

```diff
@@ -0,0 +1,390 @@
+use apollo_compiler::collections::IndexSet;
+
+use super::Key;
+use super::NamedSelection;
+use super::PathList;
+use super::Ranged;
+use super::SubSelection;
+use super::helpers::quote_if_necessary;
+use super::location::WithRange;
+
+impl PathList {
+    pub(crate) fn compute_selection_trie(&self) -> SelectionTrie {
+        let mut trie = SelectionTrie::new();
+        trie.add_path_list(self);
+        trie
+    }
+}
+
+type Ref<T> = std::sync::Arc<T>;
+
+#[derive(Debug, Eq, Clone)]
+pub(crate) struct SelectionTrie {
+    /// The top-level sub-selections of this [`SelectionTrie`].
+    selections: IndexMap<String, Ref<SelectionTrie>>,
+
+    /// Whether the path terminating at this [`SelectionTrie`] node was
+    /// explicitly added to the trie.
+    used: bool,
+
+    /// Collected as metadata but ignored by [`PartialEq`] and [`Hash`].
+    key_ranges: IndexMap<String, IndexSet<Range<usize>>>,
+}
+
+impl Display for SelectionTrie {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut need_space = false;
+
+        for (key, sub) in self.selections.iter() {
+            if need_space {
+                write!(f, " ")?;
+            }
+
+            if sub.is_empty() {
+                if sub.is_used() {
+                    write!(f, "{}", quote_if_necessary(key))?;
+                    need_space = true;
+                }
+            } else {
+                write!(f, "{} {{ {} }}", quote_if_necessary(key), sub)?;
+                need_space = true;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl PartialEq for SelectionTrie {
+    fn eq(&self, other: &Self) -> bool {
+        self.used == other.used && self.selections == other.selections
+    }
+}
+
+impl Hash for SelectionTrie {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.selections
+            .iter()
+            .fold(0, |acc, (key, sub)| {
+                let mut hasher = std::hash::DefaultHasher::default();
+                (key, sub).hash(&mut hasher);
+                acc ^ hasher.finish()
+            })
+            .hash(state);
+    }
+}
+
+impl SelectionTrie {
+    pub(crate) fn new() -> Self {
+        Self {
+            used: false,
+            selections: IndexMap::default(),
+            key_ranges: IndexMap::default(),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn new_used() -> Self {
+        Self {
+            used: true,
+            selections: IndexMap::default(),
+            key_ranges: IndexMap::default(),
+        }
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.selections.is_empty()
+    }
+
+    pub(crate) fn keys(&self) -> impl Iterator<Item = &String> {
+        self.selections.keys()
+    }
+
+    pub(crate) fn get(&self, key: impl Into<String>) -> Option<&SelectionTrie> {
+        self.selections.get(&key.into()).map(|sub| sub.as_ref())
+    }
+
+    pub(crate) fn iter(&self) -> impl Iterator<Item = (&str, &SelectionTrie)> {
```

Comment:
here you can probably just implement the iterator trait https://doc.rust-lang.org/rust-by-example/trait/iter.html

Reply:
I'm probably misunderstanding the suggestion, but I don't really want to express the iteration in terms of a stateful `next` method. Is there anything problematic about the way I've written it here?